File: server/src/main/scala/producer/message_generator/access_mode/AccessMode.scala
package producer.message_generator.access_mode

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import org.apache.pulsar.client.api.ProducerAccessMode as ProducerAccessMode

type AccessMode = ProducerAccessMode

object AccessMode:
    def fromPb(v: pb.AccessMode): ProducerAccessMode  =
        v match
            case pb.AccessMode.ACCESS_MODE_SHARED => ProducerAccessMode.Shared
            case pb.AccessMode.ACCESS_MODE_EXCLUSIVE => ProducerAccessMode.Exclusive
            case pb.AccessMode.ACCESS_MODE_WAIT_FOR_EXCLUSIVE => ProducerAccessMode.WaitForExclusive
            case pb.AccessMode.ACCESS_MODE_EXCLUSIVE_WITH_FENCING => ProducerAccessMode.ExclusiveWithFencing
            case _ => throw Exception(s"Unknown consumer access mode: $v")


    def toPb(v: ProducerAccessMode): pb.AccessMode =
        v match
            case ProducerAccessMode.Shared => pb.AccessMode.ACCESS_MODE_SHARED
            case ProducerAccessMode.Exclusive => pb.AccessMode.ACCESS_MODE_EXCLUSIVE
            case ProducerAccessMode.WaitForExclusive => pb.AccessMode.ACCESS_MODE_WAIT_FOR_EXCLUSIVE
            case ProducerAccessMode.ExclusiveWithFencing => pb.AccessMode.ACCESS_MODE_EXCLUSIVE_WITH_FENCING
            case _ => throw Exception(s"Unknown consumer access mode: $v")

File: ui/components/ui/ProducerSession/pulsar-producer-config/access-mode/access-mode.ts
import * as pb from '../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';

export type AccessMode = 'shared' | 'exclusive' | 'wait-for-exclusive' | 'exclusive-with-fencing';

export function accessModeFromPb(accessMode: pb.AccessMode): AccessMode {
  switch (accessMode) {
    case pb.AccessMode.ACCESS_MODE_SHARED: return 'shared';
    case pb.AccessMode.ACCESS_MODE_EXCLUSIVE: return 'exclusive';
    case pb.AccessMode.ACCESS_MODE_WAIT_FOR_EXCLUSIVE: return 'wait-for-exclusive';
    case pb.AccessMode.ACCESS_MODE_EXCLUSIVE_WITH_FENCING: return 'exclusive-with-fencing';
    default: throw new Error(`Unknown access mode: ${accessMode}`);
  }
}

export function accessModeToPb(accessMode: AccessMode): pb.AccessMode {
  switch (accessMode) {
    case 'shared': return pb.AccessMode.ACCESS_MODE_SHARED;
    case 'exclusive': return pb.AccessMode.ACCESS_MODE_EXCLUSIVE;
    case 'wait-for-exclusive': return pb.AccessMode.ACCESS_MODE_WAIT_FOR_EXCLUSIVE;
    case 'exclusive-with-fencing': return pb.AccessMode.ACCESS_MODE_EXCLUSIVE_WITH_FENCING;
  }
}
